packages/flutter_tools/lib/src/commands/test.dart

```diff
@@ -421,7 +418,43 @@ class TestCommand extends FlutterCommand with DeviceBasedDevelopmentArtifacts {
       webUseWasm: useWasm,
     );
 
-    final Uri? nativeAssetsJson = await nativeAssetsBuilder?.build(buildInfo);
+    final Uri projectUri = flutterProject.directory.uri;
+    final FlutterNativeAssetsBuildRunner buildRunner = FlutterNativeAssetsBuildRunnerImpl(
+      projectUri,
+      buildInfo.packageConfigPath,
+      buildInfo.packageConfig,
+      globals.fs,
+      globals.logger,
+    );
+
+    final Map<String, String> environmentDefines = <String, String>{
+      kBuildMode: buildInfo.mode.cliName,
+    };
+
+    // First perform the dart build.
+    final DartBuildResult dartBuildResult = await runFlutterSpecificDartBuild(
+        environmentDefines: environmentDefines,
+        buildRunner: buildRunner,
+        targetPlatform: TargetPlatform.tester,
+        projectUri: projectUri,
+        fileSystem: globals.fs);
+
+    Uri? nativeAssetsJson;
+    if (dartBuildResult.codeAssets.isNotEmpty) {
```

Comment:
This most likely doesn't work outside the `isolated/` directory. In g3, `package:native_assets_builder` and `package:native_assets_cli` are not imported. So their types cannot occur in any part of the code except for the Dart entry-points that are only used outside of g3.

The g3 builder is also red.

Reply:
I am aware of this. I'll look into how exactly the g3 setup is and how to massage it to make the g3 testing green.

@dcharkes do you have any other comments?

Reply:
The way in which this is set up is that `TestCompilerNativeAssetsBuilder` is defined outside `isolated/` and implemented in `isolated/` and then the implementation is threaded all the way from the entry point til where it's used. So it seems unlikely to me that `await nativeAssetsBuilder?.build(buildInfo);` and the threading of `TestCompilerNativeAssetsBuilder` can be removed.

Reply:
Ah, I'm starting to understand. The whole reason for this complexity is g3, not mockability for tests